src/librustc_mir/borrow_check.rs

```diff
@@ -846,6 +850,199 @@ impl<'c, 'b, 'a: 'b+'c, 'gcx, 'tcx: 'a> MirBorrowckCtxt<'c, 'b, 'a, 'gcx, 'tcx>
+                                    // The users have to check by themselve.
+                                    hir::MutMutable => Ok(()),
+                                }
+                            },
+                            // All other projections are owned by their base path, so mutable if
+                            // base path is mutable
+                            _ => self.is_mutable(&proj.base)
+                        }
+                    },
+                    // All other projections are owned by their base path, so mutable if
+                    // base path is mutable
+                    _ => self.is_mutable(&proj.base)
+                }
+            }
+        }
+    }
+
+    /// This lvalue owns an `&mut`, can it be used to write mutably?
+    ///
+    /// The basic idea is to enforce a unique path.
+    fn can_host_mutable_ref<'d>(&self, lvalue: &'d Lvalue<'tcx>) -> Result<(), &'d Lvalue<'tcx>> {
+        match *lvalue {
+            Lvalue::Local(..) => {
+                // Local variables are unique
+                Ok(())
+            },
+            Lvalue::Static(..) => {
+                // Static variables are not
+                Err(lvalue)
+            },
+            Lvalue::Projection(ref proj) => {
+                match proj.elem {
+                    ProjectionElem::Deref => {
+                        let base_ty = proj.base.ty(self.mir, self.tcx).to_ty(self.tcx);
+
+                        // `Box<T>` referent is unique if box is a unique spot
+                        if base_ty.is_box() {
+                            return self.can_host_mutable_ref(&proj.base);
+                        }
+
+                        // Otherwise we check the kind of deref to decide
+                        match base_ty.sty {
+                            ty::TyRef(_, tnm) => {
+                                match tnm.mutbl {
+                                    // lvalue represent an aliased location
+                                    hir::MutImmutable => Err(lvalue),
+                                    // `&mut T` is as unique as the context in which it is found
+                                    hir::MutMutable => self.can_host_mutable_ref(&proj.base),
+                                }
+                            },
+                            ty::TyRawPtr(tnm) => {
+                                match tnm.mutbl {
+                                    // `*const` is assumed to be aliasable
+                                    hir::MutImmutable => Err(lvalue),
+                                    // `*mut` can be aliased, but we leave it to user
+                                    hir::MutMutable => Ok(()),
+                                }
+                            },
+                            // Other projections are unique if the base is unique
+                            _ => self.can_host_mutable_ref(&proj.base)
```

Comment:
I'd prefer to avoid the `_` here and instead just write out all the cases explicitly. `_` makes it much harder to audit. My usual rule is something like this:

- If a new case were added to `ProjectionKind`, what is the chance that this match would be affected?

I'd say "pretty high", therefore we should avoid `_`.

Reply:
Should I use `bug!("Deref of unexpected type")` here? Like in `is_mutable()`?

Reply:
sounds good